### ISSUE TITLE
Nutrient initialisation problem

### ISSUE THREAD

**Issue by peter-devoil** [APSIMInitiative/ApsimX#8477]
### What happened?

Total carbon isn't calculated at the start of a simulation; it's zero until the end of the 1st day. So any accumulations / deltas are invalid until after then

Perhaps DoSoilOrganicMatter could be split into 2 parts - flow calculations + setting up the aggregate variables; or making the getters for aggregate variables dynamic?

### What Operating System are you on?

Linux

**Comment by hol353** [comment]
Actually even if we do what I suggest above the value of total carbon may still be wrong because some of the dependent models may not have done their calculations yet. Inter-day ordering is a pain. We could go to 'dynamic' properties that always calculate their values on demand but we know from profiling that some of those properties get called a lot each day and are expensive to call i.e. layered variables in nutrient. Ideally we want to calculate them only when necessary. Clock tries to help by having lots of events to get the timing right. Instead I wonder whether we using the declaration of [Link]s in each model to build a dependency tree and use that to order the calculations.

Sorry for the rant. I'm thinking out loud here.

**Comment by jbrider** [comment]
@hol353 Would adding a PreSimulationStart event solve your problem? All models could initialise themselves first, then be able to access other models in the StartOfSimulation event.  
It doesn't necessarily solve the ordering issue, which seems like a larger conversation, but it would solve the object not set issue.

**Comment by jbrider** [comment]
@hol353 I thought you might have, but figured if you were thinking out loud you might appreciate some external prompts.

Is it feasible to add a caching mechanism to avoid the repeated call during a single day issue? 
A simple check if the day has changed before recalculating?
It relies on the value not changing during the day and i'm not familiar enough with the carbon handling to know if that would be a problem or not?

**Comment by hol353** [comment]
Forget my previous comment above. The solution is to not report Nutrient.TotalC. Instead report:

```
[OrganicPool].C
``` 

This will already work. This will iterate through all soil organic pools and sum the 'C' property of each pool. It is still a layered variable. The downside is that perhaps it isn't as intuitive as just ```[Nutrient].TotalC```. It is listed in the report quick pick list as 'Organic Carbon'.

To do this from a manager will require a helper 'Sum' function (extension) and might look something like this:
```c#
[Link] OrganicPool[] pools;
...
pools.Sum();
```

We can then remove quite a few reports out of Nutrient (easy), then write a converter (hard) and then get Jenkins to run green (even harder). Might be a good job for @ric394 or @par456.

**Comment by peter-devoil** [comment]
Hmmm. I don't get that they are the same.

![image](https://github.com/APSIMInitiative/ApsimX/assets/1924386/a89f5db7-2a54-4634-9774-3193ecf33d73)

The use case for this was fairly basic - a crop rotation reporting on each crop/fallow transition; summing dltC and N2O at the end of each phase. The first phase (from startSimulation, C=0) is always wacky. 

Is it a matter of tightening up what we consider commence (nothing else is configured) and startSimulation (most things are present, but not necessarily configured) somehow? I think the clue is that there are two phases of initialisation described - internal and external.

**Comment by hol353** [comment]
In the end I went for a simple fix along the lines of @jbrider suggestion. I used the 'Commencing' event to do first stage initialisation and then used 'StartOfSimulation' to talk to the child models to calculate nutrient variables. This will give correct values for [Nutrient].TotalC at the beginning of the first day. 

Note: Don't report on 'StartOfSimulation'. Not all models will be correctly initialised at this time. Instead PR #8484 adds a new event called 'StartOfFirstDay' that will work.
